File: android/app/src/main/java/com/antlib/hingewave/core/EffectConfig.kt
package com.antlib.hingewave.core

import org.json.JSONObject

/** Tuned effect parameters. Mirrors core/effect.json; a unit test proves the two are identical. */
data class LaptopConfig(val armDelta: Double, val restSeconds: Double, val endAngle: Double, val armVelocity: Double)

data class InnerConfig(val clearStart: Double, val clearEnd: Double)

data class CoverConfig(val frostStart: Double, val frostEnd: Double)

data class PhoneConfig(val deadZone: Double, val inner: InnerConfig, val cover: CoverConfig)

data class SplashConfig(
    val travelSeconds: Double,
    val riseSeconds: Double,
    val stillSeconds: Double,
    val drainSeconds: Double,
    val swellHz: Double,
    val motionThreshold: Double,
)

data class EffectConfig(
    val eyeDistance: Double,
    val maxBlur: Double,
    val darkenGain: Double,
    val springHz: Double,
    val stillSeconds: Double,
    val clearSeconds: Double,
    val laptop: LaptopConfig,
    val phone: PhoneConfig,
    val splash: SplashConfig,
) {
    companion object {
        /** Values copied from core/effect.json. Change them there first, then here. */
        val DEFAULTS = EffectConfig(
            eyeDistance = 2.0,
            maxBlur = 0.036,
            darkenGain = 2.0,
            springHz = 6.0,
            stillSeconds = 2.0,
            clearSeconds = 0.6,
            laptop = LaptopConfig(armDelta = 5.0, restSeconds = 0.3, endAngle = 8.0, armVelocity = 15.0),
            phone = PhoneConfig(
                deadZone = 6.0,
                inner = InnerConfig(clearStart = 100.0, clearEnd = 174.0),
                cover = CoverConfig(frostStart = 6.0, frostEnd = 26.0),
            ),
            splash = SplashConfig(
                travelSeconds = 1.2,
                riseSeconds = 0.25,
                stillSeconds = 5.0,
                drainSeconds = 1.5,
                swellHz = 0.4,
                motionThreshold = 0.3,
            ),
        )

        fun fromJson(text: String): EffectConfig {
            val o = JSONObject(text)
            val lp = o.getJSONObject("laptop")
            val ph = o.getJSONObject("phone")
            val inner = ph.getJSONObject("inner")
            val cover = ph.getJSONObject("cover")
            val sp = o.getJSONObject("splash")
            return EffectConfig(
                eyeDistance = o.getDouble("eyeDistance"),
                maxBlur = o.getDouble("maxBlur"),
                darkenGain = o.getDouble("darkenGain"),
                springHz = o.getDouble("springHz"),
                stillSeconds = o.getDouble("stillSeconds"),
                clearSeconds = o.getDouble("clearSeconds"),
                laptop = LaptopConfig(lp.getDouble("armDelta"), lp.getDouble("restSeconds"), lp.getDouble("endAngle"), lp.getDouble("armVelocity")),
                phone = PhoneConfig(
                    deadZone = ph.getDouble("deadZone"),
                    inner = InnerConfig(inner.getDouble("clearStart"), inner.getDouble("clearEnd")),
                    cover = CoverConfig(cover.getDouble("frostStart"), cover.getDouble("frostEnd")),
                ),
                splash = SplashConfig(
                    travelSeconds = sp.getDouble("travelSeconds"),
                    riseSeconds = sp.getDouble("riseSeconds"),
                    stillSeconds = sp.getDouble("stillSeconds"),
                    drainSeconds = sp.getDouble("drainSeconds"),
                    swellHz = sp.getDouble("swellHz"),
                    motionThreshold = sp.getDouble("motionThreshold"),
                ),
            )
        }
    }
}

File: macos/Sources/HingewaveCore/EffectConfig.swift
import Foundation

/// Tuned effect parameters. Mirrors core/effect.json; a test proves the two are identical.
public struct EffectConfig: Codable, Equatable {
    public struct Laptop: Codable, Equatable {
        public var armDelta: Double
        public var restSeconds: Double
        public var endAngle: Double
        public var armVelocity: Double
        public init(armDelta: Double, restSeconds: Double, endAngle: Double, armVelocity: Double) {
            self.armDelta = armDelta
            self.restSeconds = restSeconds
            self.endAngle = endAngle
            self.armVelocity = armVelocity
        }
    }

    public struct Inner: Codable, Equatable {
        public var clearStart: Double
        public var clearEnd: Double
        public init(clearStart: Double, clearEnd: Double) {
            self.clearStart = clearStart
            self.clearEnd = clearEnd
        }
    }

    public struct Cover: Codable, Equatable {
        public var frostStart: Double
        public var frostEnd: Double
        public init(frostStart: Double, frostEnd: Double) {
            self.frostStart = frostStart
            self.frostEnd = frostEnd
        }
    }

    public struct Phone: Codable, Equatable {
        public var deadZone: Double
        public var inner: Inner
        public var cover: Cover
        public init(deadZone: Double, inner: Inner, cover: Cover) {
            self.deadZone = deadZone
            self.inner = inner
            self.cover = cover
        }
    }

    public var eyeDistance: Double
    public var maxBlur: Double
    public var darkenGain: Double
    public var springHz: Double
    public var stillSeconds: Double
    public var clearSeconds: Double
    public var laptop: Laptop
    public var phone: Phone

    public init(eyeDistance: Double, maxBlur: Double, darkenGain: Double, springHz: Double,
                stillSeconds: Double, clearSeconds: Double, laptop: Laptop, phone: Phone) {
        self.eyeDistance = eyeDistance
        self.maxBlur = maxBlur
        self.darkenGain = darkenGain
        self.springHz = springHz
        self.stillSeconds = stillSeconds
        self.clearSeconds = clearSeconds
        self.laptop = laptop
        self.phone = phone
    }

    /// Values copied from core/effect.json. Change them there first, then here.
    public static let defaults = EffectConfig(
        eyeDistance: 2.0,
        maxBlur: 0.036,
        darkenGain: 2.0,
        springHz: 6.0,
        stillSeconds: 2.0,
        clearSeconds: 0.6,
        laptop: Laptop(armDelta: 5.0, restSeconds: 0.3, endAngle: 8.0, armVelocity: 15.0),
        phone: Phone(
            deadZone: 6.0,
            inner: Inner(clearStart: 100.0, clearEnd: 174.0),
            cover: Cover(frostStart: 6.0, frostEnd: 26.0)
        )
    )

    public static func load(from url: URL) throws -> EffectConfig {
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(EffectConfig.self, from: data)
    }
}
